File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/thrift/ThriftLanguagePlugin.java
package org.jetbrains.bsp.bazel.server.sync.languages.thrift;

import ch.epfl.scala.bsp4j.BuildTarget;
import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import io.vavr.control.Option;
import java.net.URI;
import java.nio.file.Path;
import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo;
import org.jetbrains.bsp.bazel.server.bsp.utils.SourceRootGuesser;
import org.jetbrains.bsp.bazel.server.sync.BazelPathsResolver;
import org.jetbrains.bsp.bazel.server.sync.dependencytree.DependencyTree;
import org.jetbrains.bsp.bazel.server.sync.languages.LanguagePlugin;

public class ThriftLanguagePlugin extends LanguagePlugin<ThriftModule> {
  private static final String THRIFT_LIBRARY_RULE_NAME = "thrift_library";

  private final BazelPathsResolver bazelPathsResolver;

  public ThriftLanguagePlugin(BazelPathsResolver bazelPathsResolver) {
    this.bazelPathsResolver = bazelPathsResolver;
  }

  @Override
  public Set<URI> dependencySources(TargetInfo targetInfo, DependencyTree dependencyTree) {
    return HashSet.ofAll(
            dependencyTree.transitiveDependenciesWithoutRootTargets(targetInfo.getId()))
        .filter(this::isThriftLibrary)
        .flatMap(TargetInfo::getSourcesList)
        .map(bazelPathsResolver::resolveUri);
  }

  private boolean isThriftLibrary(TargetInfo target) {
    return target.getKind().equals(THRIFT_LIBRARY_RULE_NAME);
  }

  @Override
  protected void applyModuleData(ThriftModule moduleData, BuildTarget buildTarget) {
    // no actions needed
  }

  @Override
  public Option<Path> calculateSourceRoot(Path source) {
    return Option.some(SourceRootGuesser.getSourcesRoot(source));
  }
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/sync/languages/thrift/ThriftLanguagePlugin.kt
package org.jetbrains.bsp.bazel.server.sync.languages.thrift

import ch.epfl.scala.bsp4j.BuildTarget
import org.jetbrains.bsp.bazel.info.BspTargetInfo
import org.jetbrains.bsp.bazel.server.paths.BazelPathsResolver
import org.jetbrains.bsp.bazel.server.dependencygraph.DependencyGraph
import org.jetbrains.bsp.bazel.server.model.Label
import org.jetbrains.bsp.bazel.server.sync.languages.LanguagePlugin
import org.jetbrains.bsp.bazel.server.sync.languages.jvm.SourceRootGuesser
import java.net.URI
import java.nio.file.Path

class ThriftLanguagePlugin(private val bazelPathsResolver: BazelPathsResolver) :
    LanguagePlugin<ThriftModule>() {

    override fun dependencySources(
        targetInfo: BspTargetInfo.TargetInfo,
        dependencyGraph: DependencyGraph
    ): Set<URI> {
        val transitiveSourceDeps = dependencyGraph.transitiveDependenciesWithoutRootTargets(Label.parse(targetInfo.id))
            .filter(::isThriftLibrary)
            .flatMap(BspTargetInfo.TargetInfo::getSourcesList)
            .map(bazelPathsResolver::resolveUri)
            .toHashSet()

        val directSourceDeps = sourcesFromJvmTargetInfo(targetInfo)

        return transitiveSourceDeps + directSourceDeps
    }

    private fun sourcesFromJvmTargetInfo(targetInfo: BspTargetInfo.TargetInfo): HashSet<URI> =
        if (targetInfo.hasJvmTargetInfo()) {
            targetInfo
                .jvmTargetInfo
                .jarsList
                .flatMap { it.sourceJarsList }
                .map(bazelPathsResolver::resolveUri)
                .toHashSet()
        } else {
            HashSet()
        }


    private fun isThriftLibrary(target: BspTargetInfo.TargetInfo): Boolean =
        target.kind == THRIFT_LIBRARY_RULE_NAME

    protected override fun applyModuleData(moduleData: ThriftModule, buildTarget: BuildTarget) {}

    override fun calculateSourceRoot(source: Path): Path =
        SourceRootGuesser.getSourcesRoot(source)

    companion object {
        private const val THRIFT_LIBRARY_RULE_NAME = "thrift_library"
    }
}
